enforcer/src/daemon/engine.c

```diff
@@ -521,9 +544,39 @@ engine_setup(void)
     if (util_write_pidfile(engine->config->pid_filename, engine->pid) == -1) {
         hsm_close();
         ods_log_error("[%s] unable to write pid file", engine_str);
+        if (create_pipe) {
+            write(pipefd[1], "1", 1);
+            close(pipefd[1]);
+        }
         return ODS_STATUS_WRITE_PIDFILE_ERR;
     }
+    ods_log_info("[%s] enforcer started", engine_str);
+    error = hsm_open2(engine->config->repositories, hsm_prompt_pin);
+    if (error != HSM_OK) {
+        char* errorstr =  hsm_get_error(NULL);
+        if (errorstr != NULL) {
+            ods_log_error("[%s] %s", engine_str, errorstr);
+            free(errorstr);
+        } else {
+            ods_log_crit("[%s] error opening libhsm (errno %i)", engine_str,
+                error);
+        }
+        if (create_pipe) {
+            write(pipefd[1], "1", 1);
+            close(pipefd[1]);
+        }
+        return ODS_STATUS_HSM_ERR;
+    }
+    engine->need_to_reload = 0;
+    engine_start_cmdhandler(engine);
+    engine_start_workers(engine);
 
+    /* call the external start callback function */
+    start(engine);
```

Comment:
Please move these lines (6) back to the engine_run function. You're introducing an asymmetry. engine_stop_workers() etc is called at the end of engine_run(), engine_start_workers() should be called at the beginning. Since none of these calls influence writing to the pipe I see no reason to include them here. Am I missing something?